### PR TITLE
build/LazyPath: Add dirname

### LINKED ISSUE #17411: std.Build.LazyPath: add ability to return a LazyPath for the dirname
https://github.com/ziglang/zig/blob/e6590fea19e3eab94b35bfd3c36e29b53cefcaaf/lib/std/Build.zig#L1888-L1889

```zig
pub fn dirname(self: LazyPath) LazyPath {
    // ...
}
```

For union tags that have a path, `fs.path.dirname` can be used on it directly. For `GeneratedFile`, I'm thinking there could be a new field that counts how many times dirname() should be invoked when calling `getPath()`.

This must not allow a path from inside zig-cache to reach outside of zig-cache, and likewise it must not allow a path from inside a package directory to reach outside of the package directory. Every path must have a root dir that cannot be escaped from with this API.

Getting started diff:

```diff
--- a/lib/std/Build.zig
+++ b/lib/std/Build.zig
@@ -1896,6 +1896,13 @@ pub const LazyPath = union(enum) {
     /// not available until built by a build step.
     generated: *const GeneratedFile,

+    /// One of the parent directories of a file generated by an interface.
+    /// The actual path is not available until built by a build step.
+    generated_dirname: struct {
+        generated: *const GeneratedFile,
+        dirname_count: u32,
+    },
+
     /// An absolute path or a path relative to the current working directory of
     /// the build runner process.
     /// This is uncommon but used for system environment paths such as `--zig-lib-dir` which
@@ -1909,6 +1916,25 @@ pub const LazyPath = union(enum) {
         sub_path: []const u8,
     },

+    pub fn dirname(lp: LazyPath) LazyPath {
+        return switch (lp) {
+            .path => |path| .{ .path = std.fs.path.dirname(path) },
+            .generated => |generated| .{ .generated_dirname = .{
+                .generated = generated,
+                .dirname_count = 1,
+            } },
+            .generated_dirname => |gd| .{ .generated_dirname = .{
+                .generated = gd.generated,
+                .dirname_count = gd.dirname_count + 1,
+            } },
+            .cwd_relative => |path| .{ .cwd_relative = std.fs.path.dirname(path) },
+            .dependency => |dep| .{ .dependency = .{
+                .dependency = dep.dependency,
+                .sub_path = std.fs.path.dirname(dep.sub_path),
+            } },
+        };
+    }
+
     /// Returns a new file source that will have a relative path to the build root guaranteed.
     /// Asserts the parameter is not an absolute path.
     pub fn relative(path: []const u8) LazyPath {
```

### PR DESCRIPTION
**Pull request by abhinav**
Adds a variant to the LazyPath union representing a parent directory
of a generated path.

```zig
const LazyPath = union(enum) {
    generated_dirname: struct {
        generated: *const GeneratedFile,
        up: usize,
    },
    // ...
}
```

These can be constructed with the new method:

```zig
pub fn dirname(self: LazyPath) LazyPath
```

For the cases where the LazyPath is already known
(`.path`, `.cwd_relative`, and `dependency`)
this is evaluated right away.
For dirnames of generated files and their dirnames,
this is evaluated at getPath time.

dirname calls can be chained, but for safety,
they are not allowed to escape outside a root
defined for each case:

- path: This is relative to the build root,
  so dirname can't escape outside the build root.
- generated: Can't escape the zig-cache.
- cwd_relative: This can be a relative or absolute path.
  If relative, can't escape the current directory,
  and if absolute, can't go beyond root (/).
- dependency: Can't escape the dependency's root directory.

Testing:
I've included a standalone case for many of the happy cases.
I couldn't find an easy way to test the negatives, though,
because tests cannot yet expect panics.

Resolves #17411

### CODE REVIEW

**Comment by abhinav** [comment]
Ah, I assumed "zig-cache" as the basename in that test.

https://github.com/ziglang/zig/actions/runs/7324898729/job/19999209921?pr=18371#step:3:124

File: test/standalone/dirname/build.zig [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    // Generated file:
+    addTestRun(test_step, exists_in, generated.dirname(), &.{"generated.txt"});
+
+    // Generated file multiple levels:
+    addTestRun(test_step, exists_in, generated.dirname().dirname(), &.{
+        "subdir" ++ std.fs.path.sep_str ++ "generated.txt",
+    });
+
+    // Cache root:
+    addTestRun(test_step, has_basename, generated.dirname().dirname().dirname().dirname(), &.{"zig-cache"});
```
Comment (abhinav) [inline test/standalone/dirname/build.zig:42]: This should be the basename from b.cache_root.
I'll see if I can repro this failure locally and update this.

Comment (abhinav) [reply test/standalone/dirname/build.zig]: Done.

**Comment by abhinav** [comment]
Rebased following https://github.com/ziglang/zig/pull/18160

**Comment by MasterQ32** [comment]
Definitly a better solution than #16803. Thanks for taking care of that topic!

**Review by andrewrk — APPROVED** [review]
Thanks, nice work.

In the future I want to improve the implementation by moving away from absolute file paths. I took the time to write up that issue here: #18450

However, in the meantime, this is an improvement, and that issue can be a follow-up change. Thanks to your new tests, it will be easy to make sure that the feature continues to work at that time.

**Pull request merged**